bundles/org.eclipse.cdt.lsp.editor.ui/src/org/eclipse/cdt/lsp/editor/ui/clangd/CompileCommandsMonitor.java

```diff
@@ -0,0 +1,143 @@
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * 
+ * Contributors:
+ * Dominic Scharfe (COSEDA Technologies GmbH) - initial implementation
+ *******************************************************************************/
+
+package org.eclipse.cdt.lsp.editor.ui.clangd;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.eclipse.cdt.lsp.editor.ui.LspEditorUiPlugin;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.IResourceChangeEvent;
+import org.eclipse.core.resources.IResourceChangeListener;
+import org.eclipse.core.resources.IResourceDelta;
+import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.runtime.Adapters;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.jface.text.ITextViewer;
+import org.eclipse.ui.IEditorPart;
+import org.eclipse.ui.IReusableEditor;
+import org.eclipse.ui.IWorkbenchPage;
+import org.eclipse.ui.IWorkbenchWindow;
+import org.eclipse.ui.PlatformUI;
+import org.eclipse.ui.progress.UIJob;
+import org.eclipse.ui.statushandlers.StatusManager;
+
+/**
+ * Detects changes (add/delete/content) of JSON Compilation Database Format
+ * Specification files ({@value #CDBF_SPECIFICATION_JSON_FILE}) in the
+ * {@link IWorkspace workspace} and
+ * {@link CompileCommandsMonitor#refreshEditor(IEditorPart) refreshes} open
+ * editors if their {@link IEditorPart#getEditorInput()} is affected.
+ */
+public class CompileCommandsMonitor {
+	private static final String CDBF_SPECIFICATION_JSON_FILE = "compile_commands.json";
+
+	private final IResourceChangeListener listener = new IResourceChangeListener() {
+		@Override
+		public void resourceChanged(IResourceChangeEvent event) {
+			Set<IProject> projects = collectAffectedProjects(event);
+
+			if (!projects.isEmpty()) {
+				// collect all open editors which have cpp files as input and refresh them
+				Arrays.stream(PlatformUI.getWorkbench().getWorkbenchWindows()).map(IWorkbenchWindow::getPages)
+						.flatMap(Arrays::stream).map(IWorkbenchPage::getEditorReferences).flatMap(Arrays::stream)
+						.flatMap(ref -> Stream.ofNullable(ref.getEditor(false))).forEach(editor -> {
+							IFile file = Adapters.adapt(editor.getEditorInput(), IFile.class);
+
+							if (isCppFile(file) && projects.contains(file.getProject())) {
+								refreshEditor(editor, file);
+							}
+						});
+			}
+		}
+
+		private boolean isCppFile(IResource resource) {
+			try {
+				if (resource instanceof IFile && ((IFile) resource).getContentDescription() != null) {
+					String id = ((IFile) resource).getContentDescription().getContentType().getId();
+					return id.startsWith("org.eclipse.cdt.core.c") && (id.endsWith("Source") || id.endsWith("Header"));
```

Comment:
Unfortunately org.eclipse.tm4e.language_pack defines content types for c/c++ files as well. Which one will be returned by getContentType() is random. Please add a check for "lng.cpp" and "lng.c" 
E.g. ` id.equals("lng.cpp") ||` ...